reactive/flannel.py

```diff
@@ -1,254 +1,218 @@
+            'Type': 'vxlan'
+        }
+    })
+    cmd = "etcdctl "
+    cmd += "--endpoint '{0}' ".format(etcd.get_connection_string())
+    cmd += "--cert-file {0} ".format(ETCD_CERT_PATH)
+    cmd += "--key-file {0} ".format(ETCD_KEY_PATH)
+    cmd += "--ca-file {0} ".format(ETCD_CA_PATH)
+    cmd += "set /coreos.com/network/config '{0}'".format(data)
+    check_call(split(cmd))
+    set_state('flannel.network.configured')
+    remove_state('flannel.service.started')
+
+
+@when('config.changed.cidr')
+def reconfigure_network():
+    ''' Trigger the network configuration method. '''
+    remove_state('flannel.network.configured')
+
+
+@when('flannel.binaries.installed', 'flannel.service.installed',
+      'flannel.network.configured')
+@when_not('flannel.service.started')
+def start_flannel_service():
+    ''' Start the flannel service. '''
+    status_set('maintenance', 'Starting flannel service.')
+    if service_running('flannel'):
+        service_restart('flannel')
     else:
-        client = etcd.Client(host=host, port=int(port), protocol='http')
+        service_start('flannel')
+    set_state('flannel.service.started')
 
-    data = {'Network': hookenv.config('cidr'), 'Backend': {'Type': 'vxlan'}}
-    json_data = json.dumps(data)
-    client.write('/coreos.com/network/config', json_data)
 
+@when('flannel.service.started', 'cni.is-worker')
+@when_not('flannel.cni.available')
+def set_available(cni):
+    ''' Indicate to the CNI provider that we're ready. '''
+    cni.set_available()
+    set_state('flannel.cni.available')
 
-def _ingest_network_config():
-    ''' When flannel configures itself on first boot, it generates an
-    environment file (subnet.env).
 
-    return a tuple of subnet, and interface mtu'''
-
-    if not os.path.isfile('/var/run/flannel/subnet.env'):
-        # The host has not fully started, and we have no file.
-        hookenv.log('Did not find expected file: /var/run/flannel/subnet.env')
-        return
+@when('flannel.binaries.installed')
+@when_not('flannel.version.set')
+def set_flannel_version():
+    ''' Surface the currently deployed version of flannel to Juju '''
+    status_set('maintenance', 'Setting flannel version.')
+    cmd = 'flanneld -version'
+    version = check_output(split(cmd), stderr=STDOUT).decode('utf-8')
+    if version:
+        application_version_set(version.split('v')[-1].strip())
+        set_state('flannel.version.set')
 
-    with open('/var/run/flannel/subnet.env') as f:
-        flannel_config = f.readlines()
 
-    for f in flannel_config:
-        if 'FLANNEL_SUBNET' in f:
-            value = f.split('=')[-1].strip()
-            subnet = value
-        if 'FLANNEL_MTU' in f:
-            value = f.split('=')[1].strip()
-            mtu = value
-    return (subnet, mtu)
+@when('flannel.service.started')
+@when_any('cni.is-master', 'flannel.cni.available')
+def ready():
+    ''' Indicate that flannel is active. '''
+    status_set('active', 'Flannel is active.')
```

Comment:
Can we get a better detailed message than "Flannel is active"?  The existing message includes the configured subnet, which is helpful in 'at-a-glance' debugging.